### PR TITLE
Store filter settings in firebase

### PR DESCRIPTION
**Pull request by JakeThurman**
BUD-41

To Test:
- Save a filter, and make sure it persists to other views (after saving)
- Save a filter, close the app and make sure it is remembered.

### CODE REVIEW

File: Buddies/Controls/FilterSearchBar.swift [hunk 1]
```diff
@@ -65,8 +67,24 @@ class FilterSearchBar : UISearchBar, UISearchBarDelegate {
...
     
-    private func makeFilterButton(doing action: Selector, rounding corners: CACornerMask) -> UIButton {
-        let bttn = makeButton(saying: "Filter", doing: action)
+    func provideLoggedInUser(_ user: LoggedInUser?) {
+        self.user = user
+        
+        var filterSettings = user?.filterSettings ?? FilterSearchBar.defaultSettings
+        
+        // Convert to a filter state tuple
+        self.lastFilterState = (
```
Comment (noahtallen) [inline Buddies/Controls/FilterSearchBar.swift:76]: Why do we need to convert the stuff?

File: Buddies/Models/User.swift [hunk 2]
```diff
@@ -223,6 +229,7 @@ class LoggedInUser : User {
                     location: data["location"] as? GeoPoint,
                     shouldSendJoinedActivityNotification: shouldSendJoinedActivityNotification,
                     shouldSendActivitySuggestionNotification: shouldSendActivitySuggestionNotification,
+                    filterSettings: data["filter_settings"] as? [Int] ?? FilterSearchBar.defaultSettings,
```
Comment (noahtallen) [inline Buddies/Models/User.swift]: Any particular reason we use an array to store the settings? Seems like an object might be more clear

**Review by noahtallen — CHANGES_REQUESTED** [review]
Really good stuff, Jake! This appears to be working well for the test cases you outlined.

I did find one particular edge case that actually turns out to be a bit serious:
1. Navigate to topics, and then to one of the sub-activity tables.
2. Open the filter
3. Navigate to discover (keeping the filter open on topics)
4. Open the discover filter, change the settings, then save the search
5. Go back to the topics tab. The filter there will still be opened, but the changes will not be integrated.
6. If you then make changes to this open filter, the changes will _not_ persist. If you try to make a change, then save and re-open it, it will display the value from discover.
7. However, the serious part of this bug is that none of the filters on the topics tab work anymore. Everything plays nicely with each other (or so it seems) before you go through the steps above to get into this state. Once you get into this state *only* the discover filter will work - all changes to topics filters are never applied even while using them normally.

File: Buddies/Controls/FilterSearchBar.swift (hunk 1 above)
Comment (JakeThurman) [reply Buddies/Controls/FilterSearchBar.swift:76]: 1. We don't want to persist the text in the search bar
2. We can't store a tuple in Firestore so it has to change to an array

File: Buddies/Models/User.swift (hunk 2 above)
Comment (JakeThurman) [reply Buddies/Models/User.swift]: It would be a `[String : Int]` with keys. I'll change this.

File: Buddies/Controls/FilterSearchBar.swift [hunk 3]
```diff
@@ -65,8 +67,24 @@ class FilterSearchBar : UISearchBar, UISearchBarDelegate {
         ])
     }
     
-    private func makeFilterButton(doing action: Selector, rounding corners: CACornerMask) -> UIButton {
-        let bttn = makeButton(saying: "Filter", doing: action)
+    func provideLoggedInUser(_ user: LoggedInUser?) {
```
Comment (lukemeierj) [inline Buddies/Controls/FilterSearchBar.swift:70]: What is this terminology supposed to mean?  Is `provide` a common word for this context?  `setLoggedInUser` or similar seems clearer.

**Comment by noahtallen** [comment]
ps @JakeThurman is this ready for another look?

File: Buddies/Controls/FilterSearchBar.swift (hunk 3 above)
Comment (noahtallen) [reply Buddies/Controls/FilterSearchBar.swift:70]: idk, i think it's clear enough

**Comment by JakeThurman** [comment]
Not yet sorry for delaying. Life has been a little crazy. It's at the top of my priority list. For tonight/tomorrow morning.

On Mar 21, 2019 4:32 PM, Noah Allen <notifications@github.com> wrote:

ps @JakeThurman<https://github.com/JakeThurman> is this ready for another look?

File: Buddies/Controls/FilterSearchBar.swift (hunk 3 above)
Comment (JakeThurman) [reply Buddies/Controls/FilterSearchBar.swift:70]: I would say provide is common. Also, it's not so much a setter as it is something that needs to be provided.

**Comment by JakeThurman** [comment]
@noahtallen It is now ready to test

File: Buddies/Activities/ActivityTableVC.swift [hunk 4]
```diff
@@ -46,7 +46,7 @@ class ActivityTableVC: UITableViewController, FilterSearchBarDelegate {
     
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
-        fetchAndLoadActivities(for: lastSearchParam)
+        fetchAndLoadActivities(for: nil)
```
Comment (noahtallen) [inline Buddies/Activities/ActivityTableVC.swift]: @JakeThurman what's this do?

**Review by noahtallen — CHANGES_REQUESTED** [review]
So sorry, but the bug still appears to be here! To be clear, I can reproduce it by:
1. Navigate to Topics > Shopping and open the filter.
2. Leaving the filter open, navigate to discover via the tab bar
3. Open the discover tab bar and change the radius and save the filter.
4. Go back to Topics > Shopping and look at the filter.
5. The value _will_ match the new discover value
6. However, if you change the value from here, it will not persist to discover.

In other words, after doing steps 1-3, changes on discover persist to topics, but changes made to topics do not persist to discover

File: Buddies/Activities/ActivityTableVC.swift (hunk 4 above)
Comment (JakeThurman) [reply Buddies/Activities/ActivityTableVC.swift]: Tells us not to use the last params we had on this screen but to fetch the current search params.

Comment (noahtallen) [reply Buddies/Activities/ActivityTableVC.swift]: ahh cool

**Comment by JakeThurman** [comment]
@noahtallen I think I fixed it (hopefully for real this time?) lol 

(just put this on the wrong PR)

also I've just rebased! woo

**Comment by lukemeierj** [comment]
We should do another rebase from master before this is merged because #110 (in like 10 minutes will merged).  Just to make sure that we don't have the thing where both of our PRs work, but not together.

**Review by noahtallen — APPROVED** [review]
